### PR TITLE
PACT-866 Get creators through the API

### PR DESCRIPTION
**Pull request by felixgb**

### CODE REVIEW

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 1]
```diff
@@ -1750,8 +1752,8 @@ class EditSetRecordControllerSpec extends BaseControllerSpec {
     */
   private def givenCreatorsExist(returnedCreators: Seq[Creator] = Seq.empty)(implicit
     referenceDataService: ReferenceDataService
-  ): ScalaOngoingStubbing[Seq[Creator]] =
-    when(referenceDataService.getCreators).thenReturn(returnedCreators)
+  ): ScalaOngoingStubbing[IO[Seq[Creator]]] =
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1755]: Could we return a non-empty list of creators, similarly to what we do with legal statuses and places of deposit?

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 2]
```diff
@@ -1732,7 +1732,9 @@ class EditSetRecordControllerSpec extends BaseControllerSpec {
   private def givenCreatorIdsArePrepared(expectedEditSetRecord: EditSetRecord)(implicit
     editSetRecordService: EditSetRecordService
   ): ScalaOngoingStubbing[EditSetRecord] =
-    when(editSetRecordService.prepareCreatorIDs(expectedEditSetRecord)).thenReturn(expectedEditSetRecord)
+    when(
+      editSetRecordService.prepareCreatorIDs(any[Seq[Creator]], ArgumentMatchers.eq(expectedEditSetRecord))
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1736]: This wouldn't be necessary if we moved it into `EditSetRecordController`.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala [hunk 3]
```diff
@@ -31,251 +31,18 @@ import javax.inject.{ Inject, Singleton }
...
 
   def getPlacesOfDeposit(): IO[Seq[PlaceOfDeposit]] =
     apiConnector.getPlacesOfDeposit(GetPlacesOfDeposit(timestamp = timeProvider.now()))
 
   def getLegalStatuses: IO[Seq[LegalStatus]] =
     apiConnector.getLegalStatuses(GetLegalStatuses(timeProvider.now()))
 
-  def isCreatorRecognised(creatorID: String): Boolean =
-    creatorID.trim.nonEmpty && getCreators.exists(_.id == creatorID)
+  def isCreatorRecognised(creators: Seq[Creator], creatorID: String): Boolean =
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:46]: I had moved a similar function for places of deposit to the only controller where it was needed:` EditSetRecordController`. It seems more at home there.

Only a suggestion.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala [hunk 4]
```diff
@@ -22,8 +22,15 @@
...
+import uk.gov.nationalarchives.omega.editorial.models._
+import com.google.inject.ImplementedBy
 
+import javax.inject.{ Inject, Singleton }
+
+@Singleton
+class StubDataImpl @Inject() extends StubData
+
+@ImplementedBy(classOf[StubDataImpl])
 trait StubData {
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala:34]: Now sometimes we use the trait (e.g. `BaseControllerSpec`) and sometimes as an injected object (in `BaseISpec`).

I wonder if we shouldn't just always go with the latter.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala [hunk 5]
```diff
@@ -22,27 +22,31 @@
...
 import play.api.libs.json.{ Json, Reads, Writes }
 import uk.gov.nationalarchives.omega.editorial.connectors.ApiConnector.SID
 import uk.gov.nationalarchives.omega.editorial.models._
-import uk.gov.nationalarchives.omega.editorial.services.jms.ResponseBuilder.ME
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import org.typelevel.log4cats.SelfAwareStructuredLogger
 
-class ResponseBuilder[F[_] : ME : Logger] extends StubData {
+import javax.inject.{ Inject, Singleton }
+
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala:35]: Why the change from `F[_]` to `IO[_]`?

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala [hunk 6]
```diff
@@ -407,10 +425,13 @@ class EditSetRecordController @Inject() (
...
   private def prepareForDisplay(
     originalEditSetRecord: EditSetRecord,
-    placesOfDeposit: Seq[PlaceOfDeposit]
+    placesOfDeposit: Seq[PlaceOfDeposit],
+    creators: Seq[Creator]
   ): EditSetRecord =
-    Seq[EditSetRecord.Transformer](editRecordSetService.prepareCreatorIDs, preparePlaceOfDeposit(placesOfDeposit))
-      .foldLeft(originalEditSetRecord)((editSetRecord, transformer) => transformer(editSetRecord))
+    Seq[EditSetRecord.Transformer](
+      record => editRecordSetService.prepareCreatorIDs(creators, record),
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:432]: As mentioned elsewhere, I think this is better located in this controller.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala [hunk 7]
```diff
@@ -126,17 +126,16 @@ class EditSetRecordController @Inject() (
...
+          creators        <- referenceDataService.getCreators()
+        } yield BadRequest(
+          generateEditSetRecordEditView(user, editSet, record, placesOfDeposit, creators, legalStatuses, formWithErrors)
+        )
       case Left(EditSetNotFound(missingEditSetId)) =>
         BadRequest(s"Edit Set with ID [$missingEditSetId] not found")
       case Left(EditSetRecordNotFound(missingOci)) => BadRequest(s"Record with $missingOci not found")
       case Left(InvalidAction(badAction))          => BadRequest(s"$badAction is not allowed action")
       case Left(MissingAction)                     => BadRequest("This action is not allowed")
+      case Left(_)                                 => BadRequest("Unknown action")
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:138]: A good addition, but perhaps we should have a scenario that triggers it?

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala [hunk 8]
```diff
@@ -120,6 +130,8 @@ object ApiConnector {
     case object GetLegalStatuses extends SID("OSLISALS001")
     // TODO: The real SID will be provided by Adam once he figures out the schema.
     case object GetPlacesOfDeposit extends SID("OSGPOD001")
+    case object GetPersons extends SID("OSGPER001")
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:133]: Are these SIDs real, or stopgaps? 

I don't see any reference to them in the ticket.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala [hunk 9]
```diff
@@ -31,251 +31,18 @@ import javax.inject.{ Inject, Singleton }
...
-    Person("3QL", "Tate, Sir Henry", Some("1st Baronet"), Some(1819), Some(1899)),
-    Person("37K", "Uvarov, Sir Boris Petrovitch", Some("Knight"), Some(1889), Some(1970)),
-    Person("2T1", "Vane-Tempest-Stewart, Charles Stewart", Some("6th Marquess of Londonderry"), Some(1852), Some(1915)),
-    Person("4RW", "Victor Amadeus, ", Some("Duke of Savoy; King of Sardinia; King of Sicily"), Some(1666), Some(1732)),
-    Person("3GY", "Victoria, ", Some("Queen of Great Britain and Ireland"), Some(1819), Some(1901))
-  )
-
-  def getCreators: Seq[Creator] =
-    (getCorporateBodies.flatMap(Creator.from) ++ getPersons.flatMap(Creator.from)).sortBy(_.name)
+  def getCreators(): IO[Seq[Creator]] =
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:34]: A deviation from the requirements, but this is more correct.

**Review by tna-erasmos — COMMENTED** [review]
Nothing substantial except for the apparently unnecessary support for JSON for the `Creator`.

**Comment by tna-erasmos** [comment]
Details on my superficial (but probably adequate) QA [here](https://national-archives.atlassian.net/browse/PACT-866?focusedCommentId=45055).

File: build.sbt [hunk 10]
```diff
@@ -149,6 +149,8 @@ lazy val root = Project("ctd-omega-editorial-frontend", file("."))
       "org.typelevel"          %% "log4cats-core"                 % "2.5.0",
       "org.typelevel"          %% "log4cats-slf4j"                % "2.5.0",
       "org.webjars.npm"         % "govuk-frontend"                % "4.3.1",
+      "com.beachape"           %% "enumeratum-play-json"          % "1.7.2",
```
[the remark this answers is not in the document]
Comment (felixgb) [reply build.sbt:152]: It was for the creator json format, which is unused now. I'll delete these.

File: test/support/BaseSpec.scala [hunk 11]
```diff
@@ -49,7 +49,6 @@ class BaseSpec
     with ApiConnectorAssertions {
 
   val user: User = User("dummy user")
-  val testReferenceDataService: TestReferenceDataService = app.injector.instanceOf[TestReferenceDataService]
   val editSetRecordService: EditSetRecordService = app.injector.instanceOf[EditSetRecordService]
   val editSetService: EditSetService = app.injector.instanceOf[EditSetService]
   val legalStatuses: Seq[LegalStatus] = Seq(
```
[the remark this answers is not in the document]
Comment (felixgb) [reply test/support/BaseSpec.scala:54]: There are different sets of test data here. I can make this class use the reference data, but I imagine the tests will need refactoring. Do you think I should do that in this ticket?

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 1 above)
Comment (felixgb) [reply test/controllers/EditSetRecordControllerSpec.scala:1755]: I've tried changing this to a list of creators, but mockito doesn't like it:
```
[info]  - this invocation of 'apply' method:
[info]     editSetRecordEdit.apply(
[info]     User("1234"),
[info]     "COAL 80 Sample",
[info]     "edit-set.record.edit.title",
[info]     EditSetRecord("COAL 80/80/8", "COAL.2022.V8RJW.P", "Bedlington Colliery, Newcastle Upon Tyne. Photograph depicting: view of pithead baths. (H)", "1969", "", "", "1", "1", "1969", "31", "12", "1969", "ref.1", "1", "", "", "", List(), List(), List("48N", "46F", "8R6")),
[info]     List(LegalStatus("ref.1", "Public Record(s)"), LegalStatus("ref.2", "Not Public Records"), LegalStatus("ref.3", "Public Records unless otherwise Stated"), LegalStatus("ref.4", "Welsh Public Record(s)")),
[info]     List(PlaceOfDeposit("1", "The National Archives, Kew"), PlaceOfDeposit("2", "British Museum, Department of Libraries and Archives"), PlaceOfDeposit("3", "British Library, National Sound Archive")),
[info]     List(Creator(uk.gov.nationalarchives.omega.editorial.models.Creator$CreatorType$CorporateBody$@2cd16f3f, "RR6", "100th (Gordon Highlanders) Regiment of Foot", Some(1794), Some(1894)), Creator(uk.gov.nationalarchives.omega.editorial.models.Creator$CreatorType$CorporateBody$@2cd16f3f, "JS8", "BBC", None, None), Creator(uk.gov.nationalarchives.omega.editorial.models.Creator$CreatorType$CorporateBody$@2cd16f3f, "S2", "The National Archives", Some(2003), None), Creator(uk.gov.nationalarchives.omega.editorial.models.Creator$CreatorType$Person$@4e9a5f95, "3FH", "Dainton, Sir Frederick Sydney", Some(1914), Some(1997)), Creator(uk.gov.nationalarchives.omega.editorial.models.Creator$CreatorType$Person$@4e9a5f95, "39K", "Cannon, John Francis Michael", Some(1930), None)),
[info]     Form(play.api.data.ObjectMapping16@36fbf5a2, HashMap("custodial-history" -> "", "end-date-month" -> "12", "covering-dates" -> "1969", "former-reference-department" -> "", "oci" -> "COAL.2022.V8RJW.P", "creator-ids[0]" -> "92W", "legal-status-id" -> "ref.1", "former-reference-pro" -> "", "note" -> "", "end-date-day" -> "31", "end-date-year" -> "1969", "creator-ids[1]" -> "48N", "action" -> "removeLastCreator", "start-date-month" -> "1", "start-date-year" -> "1969", "scope-and-content" -> "Bedlington Colliery, Newcastle Upon Tyne. Photograph depicting: view of pithead baths. (H)", "background" -> "", "place-of-deposit-id" -> "1", "start-date-day" -> "1"), List(), Some(EditSetRecordFormValues("Bedlington Colliery, Newcastle Upon Tyne. Photograph depicting: view of pithead baths. (H)", "1969", "", "", "1", "1", "1969", "31", "12", "1969", "ref.1", "1", "", "", "", List("92W", "48N", "8R6")))),
[info]     MessagesImpl(Lang(en_GB), play.api.i18n.DefaultMessagesApi@7dee08f8),
[info]     GET /
[info] );
[info]     -> at uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.generateEditSetRecordEditView(EditSetRecordController.scala:465)
[info]  - has following stubbing(s) with different arguments:
[info]     1. editSetRecordEdit.apply(
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null,
[info]     null
[info] );
[info]       -> at controllers.EditSetRecordControllerSpec.givenEditViewIsGenerated(EditSetRecordControllerSpec.scala:1779)
[info] Typically, stubbing argument mismatch indicates user mistake when writing tests.
[info] Mockito fails early so that you can debug potential problem easily.
[info] However, there are legit scenarios when this exception generates false negative signal:
[info]   - stubbing the same method multiple times using 'given().will()' or 'when().then()' API
[info]     Please use 'will().given()' or 'doReturn().when()' API for stubbing.
[info]   - stubbed method is intentionally invoked with different arguments by code under test
[info]     Please use default or 'silent' JUnit Rule (equivalent of Strictness.LENIENT).
[info] For more information see javadoc for PotentialStubbingProblem class.
[info]         at uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.generateEditSetRecordEditView(EditSetRecordController.scala:465)
[info]         at uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.$anonfun$removeLastCreator$3(EditSetRecordController.scala:290)
[info]         at map @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.$anonfun$removeLastCreator$2(EditSetRecordController.scala:280)
[info]         at flatMap @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.$anonfun$removeLastCreator$1(EditSetRecordController.scala:279)
[info]         at flatMap @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.removeLastCreator(EditSetRecordController.scala:278)
[info]         at flatMap @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.submit(EditSetRecordController.scala:119)
[info]         at map @ uk.gov.nationalarchives.omega.editorial.controllers.BaseAppController.findEditSetRecord(BaseAppController.scala:53)
[info]         at map @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.$anonfun$generateResult$2(EditSetRecordController.scala:151)
[info]         at map @ uk.gov.nationalarchives.omega.editorial.controllers.BaseAppController.findEditSet(BaseAppController.scala:48)
[info]         at flatMap @ uk.gov.nationalarchives.omega.editorial.controllers.EditSetRecordController.$anonfun$generateResult$1(EditSetRecordController.scala:150)
[info]         ...

```
I've spent days debugging mockito errors before, so I'm going to come back to this one after fixing all the other bugs

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 2 above)
Comment (felixgb) [reply test/controllers/EditSetRecordControllerSpec.scala:1736]: Moving `prepareCreatorIDs` into the controller has broken the tests - mockito fails at runtime in a way I do not understand.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala (hunk 3 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:46]: OK, moving this method also creates a stubbing exception.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala (hunk 4 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala:34]: Some classes are using guice for dependency injection, and some are using trait inheritance hierarchies. IMHO it would be a major refactor to use guice everywhere. Most tests are also using traits.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala (hunk 5 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala:35]: In short - we were not using it, and I need to be able to inject this class with Guice. Type parameters make that a lot harder.

We're not using any other effect type here instead of IO. If we were using generic effects then I would consider keeping it, as it is, I think it's unnecessarily complicated and doesn't fit the style of the rest of this application (using Guice)

File: app/uk/gov/nationalarchives/omega/editorial/models/Creator.scala [hunk 12]
```diff
@@ -42,22 +51,29 @@ case class Creator(creatorType: CreatorType, id: String, name: String, startYear
 
 object Creator {
 
-  sealed trait CreatorType
+  implicit val format: Format[Creator] = Json.format[Creator]
```
[the remark this answers is not in the document]
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/models/Creator.scala:54]: I deleted the format, it wasn't being used. When I started this ticket I had the creator being retrieved by the API connector.

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala (hunk 8 above)
Comment (felixgb) [reply app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:133]: These SID are made up, I can't find any real SIDs for these.

File: build.sbt (hunk 10 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply build.sbt:152]: OK.

File: test/support/BaseSpec.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/support/BaseSpec.scala:54]: No, not if it would require substantial extra work.

File: test/support/BaseSpec.scala [hunk 13]
```diff
@@ -63,7 +62,16 @@ class BaseSpec
     PlaceOfDeposit("2", "British Museum, Department of Libraries and Archives"),
     PlaceOfDeposit("3", "British Library, National Sound Archive")
   )
-  val allCreators: Seq[Creator] = testReferenceDataService.getCreators
+  val allCreators: Seq[Creator] = Seq(
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/support/BaseSpec.scala:65]: Yes; only for more transparency, mind you. At this level, we don't need to know of the sausage is made.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala (hunk 3 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala:46]: Let's look into it; will likely be easy to resolve.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala (hunk 4 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/jms/StubData.scala:34]: In that case, best to leave it.

And it may be more flexible this way.

File: app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala (hunk 5 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/jms/ResponseBuilder.scala:35]: That's what I thought: we don't need the flexibility.

I think it's a good step.

File: app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala (hunk 8 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/connectors/ApiConnector.scala:133]: That's what I thought. In that case, maybe copy the TODO above, to each? Or something like: the following SIDs are made up, but will be replaced ...

File: build.sbt (hunk 10 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply build.sbt]: Confirmed.

File: test/support/BaseSpec.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/support/BaseSpec.scala:66]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 1 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala:1749]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 2 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala (hunk 3 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/services/ReferenceDataService.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/models/Creator.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/Creator.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala (hunk 7 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala:138]: We spoke about this, and it would be more trouble than it's worth - at least how it's currently implemented.

I'm OK with postponing this.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala (hunk 6 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetRecordController.scala]: Confirmed.

**Review by tna-erasmos — APPROVED** [review]
LGTM

**Review by rwalpole — APPROVED** [review]
